pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+      int targetNumReplicaGroup, int targetNumInstancesPerPartition, List<String> addedServers,
+      List<String> removedServers) {
+    boolean sameNumInstancesPerPartition = oldNumInstancesPerPartition == targetNumInstancesPerPartition;
+    boolean sameNumReplicaGroup = oldNumReplicaGroup == targetNumReplicaGroup;
+    boolean isAddedServersSizeZero = addedServers.size() == 0;
+    boolean isRemovedServersSizeZero = removedServers.size() == 0;
+
+    if (sameNumInstancesPerPartition && sameNumReplicaGroup && addedServers.size() == removedServers.size()) {
+      return ReplicaGroupRebalanceType.REPLACE;
+    } else if (sameNumInstancesPerPartition) {
+      if (oldNumReplicaGroup < targetNumReplicaGroup && !isAddedServersSizeZero && isRemovedServersSizeZero
+          && addedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.ADD_REPLICA_GROUP;
+      } else if (oldNumReplicaGroup > targetNumReplicaGroup && isAddedServersSizeZero && !isRemovedServersSizeZero
+          && removedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_REPLICA_GROUP;
+      }
+    } else if (sameNumReplicaGroup) {
+      if (oldNumInstancesPerPartition < targetNumInstancesPerPartition && isRemovedServersSizeZero
+          && !isAddedServersSizeZero && addedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.ADD_SERVER;
+      } else if (oldNumInstancesPerPartition > targetNumInstancesPerPartition && !isRemovedServersSizeZero
+          && isAddedServersSizeZero && removedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_SERVER;
+      }
+    }
+    return ReplicaGroupRebalanceType.UNSUPPORTED;
+  }
+
+  /**
+   * Rebalances segments for a replica group table without update
+   * @param idealState old idealstate
+   * @param tableConfig a table config
+   * @param replicaGroupPartitionAssignment a replica group partition assignment
+   * @param rebalanceGroupBootstrap bootstrap mode
+   * @return a rebalanced idealstate
+   */
+  private IdealState rebalanceSegments(IdealState idealState, TableConfig tableConfig,
+      ReplicaGroupPartitionAssignment replicaGroupPartitionAssignment, boolean rebalanceGroupBootstrap) {
+    Map<String, Map<String, String>> segmentToServerMapping = idealState.getRecord().getMapFields();
+    Map<String, LinkedList<String>> serverToSegments =
+        buildServerToSegmentMapping(segmentToServerMapping, rebalanceGroupBootstrap);
+
+    // Compute added and removed servers
+    List<String> oldServerInstances = new ArrayList<>(serverToSegments.keySet());
+    List<String> serverInstances = replicaGroupPartitionAssignment.getAllInstances();
+
+    List<String> addedServers = new ArrayList<>();
+    List<String> removedServers = new ArrayList<>();
+    for (String server : oldServerInstances) {
```

Comment:
use removeAll() like in other method reviewed

Reply:
fixed